presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineSemiJoinDistributionType.java

```diff
@@ -0,0 +1,162 @@
+import com.facebook.presto.cost.TaskCountEstimator;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.google.common.collect.Ordering;
+import io.airlift.units.DataSize;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.SystemSessionProperties.getJoinMaxBroadcastTableSize;
+import static com.facebook.presto.cost.CostCalculatorWithEstimatedExchanges.calculateJoinCostWithoutOutput;
+import static com.facebook.presto.sql.planner.plan.Patterns.semiJoin;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.REPLICATED;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This rule must run after the distribution type has already been set for delete queries,
+ * since semi joins in delete queries must be replicated.
+ * Once we have better pattern matching, we can fold that optimizer into this one.
+ */
+public class DetermineSemiJoinDistributionType
+        implements Rule<SemiJoinNode>
+{
+    private final TaskCountEstimator taskCountEstimator;
+    private final CostComparator costComparator;
+
+    private static final Pattern<SemiJoinNode> PATTERN = semiJoin().matching(semiJoin -> !semiJoin.getDistributionType().isPresent());
+
+    public DetermineSemiJoinDistributionType(CostComparator costComparator, TaskCountEstimator taskCountEstimator)
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    @Override
+    public Pattern<SemiJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SemiJoinNode semiJoinNode, Captures captures, Context context)
+    {
+        JoinDistributionType joinDistributionType = getJoinDistributionType(context.getSession());
+        switch (joinDistributionType) {
+            case AUTOMATIC:
+                return Result.ofPlanNode(getCostBasedDistributionType(semiJoinNode, context));
+            case PARTITIONED:
+                return Result.ofPlanNode(semiJoinNode.withDistributionType(PARTITIONED));
+            case BROADCAST:
+                return Result.ofPlanNode(semiJoinNode.withDistributionType(REPLICATED));
+            default:
+                throw new UnsupportedOperationException("Unknown join_distribution_type: " + joinDistributionType);
```

Comment:
`UnsupportedOperationException` should be used when the entire operation (method) is not supported. You can use here either `IllegalArgumentException` or `VerifyException`

Reply:
ok. There are other places where we throw UnsupportedOperationException for switch statements, but I see that it doesn't quite fit.  I'll change it here.

Reply:
Yeah, indeed. Feel free to keep it as is then.